### PR TITLE
Vol→vol self-attention block post surf→vol xattn (decoder pattern)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The surf→vol cross-attention (PR #823) lets volume tokens query surface geometry — a proven broad-spectrum win. But after xattn, each vol token has only received a surface-conditioning signal; the vol tokens cannot yet communicate with **each other**. Volumetric pressure in CFD has inherently non-local spatial structure (wake propagation, pressure gradients, stagnation zones) that requires vol-to-vol communication, not just surface conditioning.

This PR adds a **vol-only self-attention block** immediately after the surf→vol xattn residual update, before the volume regression head. The block lets vol tokens spatially integrate the surface-conditioning signal they just received. This is architecturally distinct from:
- PR #884 / #890 (two-layer xattn — K/V gradient backflow NEGATIVE): we are NOT adding another surf→vol xattn, only a vol→vol self-attention
- PR #891 (post-xattn FFN NEGATIVE): we are NOT just adding an FFN, but adding full spatial attention among vol tokens

**Why this matters for test_vol_pressure OOD:** The 4 OOD test cases likely produce unusual pressure fields in specific spatial regions. A vol-self-attention block after xattn allows vol tokens near the anomalous geometry (which just received surface conditioning signals via xattn) to spatially communicate and form a coherent volume pressure field. An FFN alone cannot do this — it processes each vol token independently.

**Implementation:** Add one `nn.MultiheadAttention(embed_dim=512, num_heads=4, batch_first=True)` + `nn.LayerNorm(512)` + `nn.Linear(512, 2048) + GELU + nn.Linear(2048, 512)` block that operates on vol_hidden **after** the surf→vol xattn output. Zero-init the out_proj of the attention and the second linear of the FFN. Residual connections; pre-norm (LayerNorm before attention, LayerNorm before FFN).

**References:** This is the standard "decoder block" pattern from the original Transformer (Vaswani 2017) — encoder cross-attention followed by self-attention + FFN on the decoder side. We are using it here as a "volume decoder refinement block" after xattn conditioning.

## Instructions

1. **Add a new flag** `--use-vol-self-attn-block` (bool, default False) to `train.py`

2. **Add a `VolSelfAttnBlock` module** in `train.py` (or the model file):
   ```python
   class VolSelfAttnBlock(nn.Module):
       def __init__(self, hidden_dim: int, num_heads: int, mlp_ratio: float = 4.0):
           super().__init__()
           self.norm1 = nn.LayerNorm(hidden_dim)
           self.self_attn = nn.MultiheadAttention(
               embed_dim=hidden_dim, num_heads=num_heads,
               batch_first=True, dropout=0.0
           )
           self.norm2 = nn.LayerNorm(hidden_dim)
           self.ffn = nn.Sequential(
               nn.Linear(hidden_dim, int(hidden_dim * mlp_ratio)),
               nn.GELU(),
               nn.Linear(int(hidden_dim * mlp_ratio), hidden_dim),
           )
           # zero-init output projections for identity-at-init
           nn.init.zeros_(self.self_attn.out_proj.weight)
           nn.init.zeros_(self.self_attn.out_proj.bias)
           nn.init.zeros_(self.ffn[-1].weight)
           nn.init.zeros_(self.ffn[-1].bias)

       def forward(self, x):  # x: (B, N_vol, hidden_dim)
           # Pre-norm self-attention
           residual = x
           x_norm = self.norm1(x)
           attn_out, _ = self.self_attn(x_norm, x_norm, x_norm, need_weights=False)
           x = residual + attn_out
           # Pre-norm FFN
           residual = x
           x = residual + self.ffn(self.norm2(x))
           return x
   ```

3. **Insert the block** after the surf→vol xattn residual update, before the vol regression head:
   ```python
   # Existing: vol_hidden = vol_hidden + xattn_out  (residual from surf→vol xattn)
   # ADD: if self.vol_self_attn_block is not None:
   #          vol_hidden = self.vol_self_attn_block(vol_hidden)
   # Then: vol_out = self.vol_head(vol_hidden)
   ```

4. **DDP constraint**: Add `find_unused_parameters=True` on the DDP wrapper (already required for `--use-surf-to-vol-xattn` — ensure it also covers this new module).

5. **Run the 4-epoch screen** with the standard SOTA config + `--use-surf-to-vol-xattn --use-vol-self-attn-block`:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-self-attn-block \
  --wandb-group edward-vol-self-attn-post-xattn \
  --wandb-name edward/vol-self-attn-screen
```

6. **Kill gates** (report val_abupt at each epoch boundary):
   - EP1 (step ~10,864): must be **≤30%**
   - EP2 (step ~21,728): must be **≤16%**
   - EP3 (step ~32,592): must be **≤8.0%**
   - EP4 (step ~38,028 or timeout): must be **<6.9%** to proceed to Phase 2

7. **If Phase 1 passes** (EP4 < 6.9%): run the full 13-epoch schedule with the same flags, replacing `--epochs 4 --lr-cosine-t-max 13` with `--epochs 13 --lr-cosine-t-max 13` and using the 13-epoch vol-curriculum `--vol-points-schedule "0:16384:3:32768:6:49152:9:65536"`.

## Baseline to Beat

| Metric | Single-model SOTA (PR #823) | Target |
|---|---|---|
| val_abupt | **6.4407%** | < 6.4407% |
| test_abupt | 7.6992% | — |
| test_vol_pressure | 11.6704% | ≤ 11.0% (weak win) |
| val_vol_pressure | 3.8557% | — |

W&B run: `ghh0s4ne` (group `nezuko-surf-vol-xattn`)

**Note:** `--use-surf-to-vol-xattn` is required and must come before `--use-vol-self-attn-block`. The `find_unused_parameters=True` DDP flag must be active when either of these is used.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Edward — your pod is healthy and the experiment spec is fully defined in the PR body. Please implement the VolSelfAttnBlock and launch the 4-epoch screen now. The run command is in the PR body. Gate thresholds: EP1 ≤30%, EP2 ≤16%, EP3 ≤8.0%, EP4 <6.9% (val_abupt). Report back after EP1 completes. Current baseline to beat: val_abupt=6.4407%.

**Comment by morganmcg1** [comment]
STUDENT edward: EP1 update — Phase 1 4-epoch screen passes the EP1 kill gate, but only barely.

## EP1 Results (step 10864)

| Metric | This run | Baseline (#823 PR `ghh0s4ne`) | Δ |
|---|---|---|---|
| val_abupt | **29.6892%** | 28.6344% | +1.06 pp ❌ |
| val_vol_pressure | 14.9469% | 17.7870% | -2.84 pp ✅ |
| val_surf_pressure | 24.3410% | 21.8521% | +2.49 pp ❌ |
| val_loss | 0.10513 | — | — |
| epoch_time | 6066 s (~101 min) | — | — |

**EP1 kill gate (≤30%)**: ✅ PASS (margin 0.31 pp)

## Status
- All 8 ranks running (`run-id rank0 = nmvw5t2d`, group `edward-vol-self-attn-post-xattn`).
- Currently mid-EP2 (step 770/5435 of EP2, ~14% through). ETA EP4 ≈ 21:30 UTC.
- DDP wrapper with `find_unused_parameters=True` confirmed active, no unused params warning at init (expected since vol_self_attn_block is identity but gradients flow through it).

## Quick read
- Volume-pressure side **already 2.84 pp better than baseline at EP1** — promising signal that vol→vol self-attention helps volumetric coherence even at this early stage.
- Surface and abupt are slightly slower at EP1 — likely because the zero-init block divides gradient signal across more params; expected to recover by EP2.
- Tight margin on the EP1 gate suggests this experiment will be a close call at EP4 (baseline EP4 = 6.81%, gate <6.9%). Will report EP2 when it lands.

Continuing the run; will report back at EP2 (gate ≤16%, baseline 8.15%).

**Comment by morganmcg1** [comment]
STUDENT edward: EP2 update — gate passed but trajectory below baseline.

## EP2 Results (step 16299)

| Metric | This run (EP2) | Baseline #823 (EP2) | Δ |
|---|---|---|---|
| val_abupt | **11.1636%** | 8.1482% | +3.02 pp ❌ |
| val_surf_pressure | 7.4975% | 5.4090% | +2.09 pp ❌ |
| val_vol_pressure | 8.0258% | 4.9559% | +3.07 pp ❌ |
| val_wall_shear | 12.1756% | 9.1903% | +2.99 pp ❌ |
| val_loss | 0.01528 | 0.00816 | +0.0071 ❌ |
| epoch_time | 4370.8 s (~73 min) | — | — |

**EP2 kill gate (≤16%)**: ✅ PASS (margin 4.84 pp), but well below baseline pace.

## Trajectory analysis
- EP1 vol_pressure was promising (14.95% vs baseline 17.79%, -2.84 pp), but the curriculum jumping from 16k→32k vol points at EP1 boundary appears to have erased that advantage.
- EP2 vol_pressure is now +3.07 pp worse than baseline. The vol-self-attn block is converging slower than baseline across the board.
- Baseline went EP1→EP2: 28.63%→8.15% (Δ 20.48 pp); ours went 29.69%→11.16% (Δ 18.53 pp). Ours is ~2 pp slower at closing the gap.

## EP3/EP4 forecast
- Baseline EP3=7.12%, EP4=6.81%. Ours starts EP3 with a 3 pp deficit.
- Naively projecting current pace, EP3 likely lands ~9-10% (close to 8% kill gate, may fail), EP4 likely ~8-9% (would fail <6.9% gate).
- Will continue and report EP3 metrics. Kill gate is ≤8.0% — this run is at risk.

## Status
- Run-id rank0 = `nmvw5t2d`, group `edward-vol-self-attn-post-xattn`.
- Currently early-EP3 (~8% complete). ETA EP3 ≈ 19:00 UTC, EP4 ≈ 20:30 UTC.
- All 8 ranks still running, no NaN/OOM events.

Continuing run; will report at EP3 (gate ≤8.0%).

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["nmvw5t2d"],"primary_metric":{"name":"val_abupt","value":8.2271},"test_metric":{"name":"test_abupt","value":null,"note":"run terminated mid-EP4 before final test eval"}}

## Final Results — EP3 kill gate FAILED (terminal)

### EP-by-EP trajectory vs PR #823 baseline

| Metric | EP1 | EP2 | EP3 | Baseline EP3 | Δ EP3 |
|---|---:|---:|---:|---:|---:|
| val_abupt | 29.6892% | 11.1636% | **8.2271%** | 7.12% | +1.11 pp ❌ |
| val_surf_pressure_rel_l2 | 24.3410% | 7.4975% | 5.4071% | — | — |
| val_vol_pressure_rel_l2 | 14.9469% | 8.0258% | 5.2888% | — | — |
| val_wall_shear_rel_l2 | 33.5162% | 12.1756% | 9.1776% | — | — |
| val_loss | 0.10513 | 0.01528 | 0.00833 | — | — |
| epoch_time_s | 6065.8 | 4370.8 | 3983.0 | — | — |

(per-channel MAE at EP3: surf=0.01531, vol=10.2965, wall=0.08418)

### Kill gate verdict

| Epoch | Kill gate | Result | Status |
|---|---|---:|:---|
| EP1 | ≤30% | 29.6892% | ✅ PASS (margin 0.31 pp) |
| EP2 | ≤16% | 11.1636% | ✅ PASS (margin 4.84 pp) |
| EP3 | **≤8.0%** | **8.2271%** | **❌ FAIL** (over by 0.23 pp) |
| EP4 | <6.9% | killed at step 282/2720 (~10%) | — (ext. SIGTERM at 18:59:16 UTC, ~252 min runtime) |

EP3 missed the gate. The trajectory was already 3.0 pp behind baseline at EP2 and held a 1.1 pp deficit through EP3. EP4 needed to drop val_abupt from 8.23% → <6.9% (a 1.34 pp single-epoch decline) — far beyond the baseline's own EP3→EP4 pace of 0.31 pp. No realistic path to the EP4 gate. **Marking terminal.**

### What happened

The vol-self-attention block converged consistently slower than baseline:
- **EP1 (+1.06 pp)**: zero-init makes it near-identity; this is initialization noise.
- **EP2 (+3.02 pp)**: the new block is now actively dragging on convergence — this was the warning sign.
- **EP3 (+1.11 pp)**: partial recovery as the block specializes, but still well off pace.

### Why it likely failed

1. **Optimizer warm-up cost too large for 4-epoch budget.** The block adds two layer norms + MHA(out_proj zero-init) + MLP(last linear zero-init). With Lion at lr=9e-5 and 1-epoch warmup, the new params need time to develop a useful update direction. The baseline's weights are already near-mature at EP3 (7.12%), but ours are still chasing that same stationary point with extra free parameters. The trajectory shape (peaks divergence at EP2 then narrows by EP3) is consistent with a trainable-but-immature module — not a broken one.
2. **Param count up ~3M** (20.14M total reported in W&B vs ~17.4M baseline single-block surf→vol setup). All extra capacity is in the volume-decoder side, but the val_vol_pressure metric at EP3 (5.29%) is ~1.5 pp behind baseline EP3 vol_pressure (~3.8%). The extra capacity isn't paying off here.
3. **`find_unused_parameters=True` overhead is real but not the root cause.** DDP confirmed no unused params at init — the block sits in the gradient path every step.
4. **No NaN/grad blow-up observed.** Slopes (val/slope/...) all negative — module is learning, just too slowly.

### Observation: post-xattn capacity additions are 0-for-3

| Idea | PR | Outcome |
|---|---|---|
| Two-layer surf→vol xattn (K/V backflow) | #884, #890 | NEGATIVE |
| Post-xattn FFN | #891 | NEGATIVE |
| Post-xattn vol-self-attn (this PR) | #906 | NEGATIVE |

The advisor's framing was right that this PR is architecturally distinct from the others, but the 4-epoch screen still rejects all three. The bottleneck for the screen budget appears not to be post-xattn refinement.

### Suggested follow-ups

1. **Don't repeat post-xattn capacity additions on the 4-epoch screen format.** The pattern is settled — three orthogonal additions all fail similarly.
2. **If revisiting in a 13-epoch run**, try a small non-zero init scale on the FFN's last linear (e.g. `gain=1e-4`) to give the optimizer non-zero gradient signal earlier. Pure zero-init may be too cold a start for a 4-epoch budget.
3. **Vol-self-attn BEFORE the surf→vol xattn** is structurally different — let vol tokens pre-aggregate among themselves before being conditioned on surface. This puts the new module earlier in the gradient path and gives it longer to mature, but reorders the encoder so it isn't a clean comparison to baseline.
4. **Increase vol_head depth/width** (e.g. wider MLP or 2-layer head) — same motivation (more vol-side compute) but smaller architectural perturbation that doesn't change gradient routing through the encoder.
5. **Cross-cutting:** the `val_vol_pressure` test/val gap (3.6% val vs 11.3% test in BASELINE.md ensemble row) is the standing systematic issue — vol-side improvements that target test OOD generalization (e.g. data-augmentation, distance-weighted vol sampling) may pay better than capacity additions.

### Run details

- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/nmvw5t2d (group: `edward-vol-self-attn-post-xattn`)
- Run state: `crashed` (external SIGTERM at 18:59:16 UTC during EP4 step 282/2720; W&B runtime 4h 11m 44s, well below the 270-min `train_timeout_minutes`. No NaN/OOM in logs; no `train timeout` log line. Likely pod-level or harness-level external kill since the pod-level `SENPAI_TIMEOUT_MINUTES=360` was also not breached.)
- Param count: 20,139,609 (20.14M)
- Peak GPU memory (reserved, pre-clear): 63.0 GB / 80 GB

### Command run

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-vol-self-attn-block \
  --wandb-group edward-vol-self-attn-post-xattn \
  --wandb-name edward/vol-self-attn-screen
```

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #906 because ADVISOR: Closing as a dead end. Run nmvw5t2d failed the EP3 kill gate (8.2271% > 8.0%) before the external SIGTERM at step 282 — the experiment was already terminal. No per-channel signal that vol-self-attn after the surf→vol xattn helped on the val/test gap.

This is the third negative result in the same pattern:
- PR #884/#890: two-layer xattn (failed)
- PR #891: post-xattn FFN (failed)
- PR #906: post-xattn vol-self-attn (failed)

The signal is now strong: post-xattn capacity additions to the volume branch don't move the val/test gap on volume_pressure. We'll redirect this thread.

Edward, thanks for the clean implementation (zero-init out_proj, find_unused_parameters=True, healthy training curve) — execution was correct, the hypothesis just doesn't pan out. New assignment incoming..
